bitnami/keycloak/templates/servicemonitor.yaml

```diff
@@ -20,25 +20,27 @@ spec:
   jobLabel: {{ .Values.metrics.serviceMonitor.jobLabel }}
   {{- end }}
   endpoints:
-    - port: http
-      {{- if .Values.metrics.serviceMonitor.path }}
-      path: {{ .Values.metrics.serviceMonitor.path }}
+    {{- $endpoints := ternary (.Values.metrics.serviceMonitor.endpoints) (list (dict "path" .Values.metrics.serviceMonitor.path)) (empty .Values.metrics.serviceMonitor.path) }}
+    {{- range $endpoints }}
+    {{- $endpoint := merge . $.Values.metrics.serviceMonitor }}
```

Comment:
@maxnitze, this is a rather big issue as this merge here creates a reference cycle.
Merge modifies the "dest" argument, which in this case is potentially a dict from the `serviceMonitor.endpoints` list. Since we're merging with the `serviceMonitor` object itself, we end up inserting a reference to `serviceMonitor.endpoints` into every item. In other words, we can do `$endpoint.endpoints[0].endpoints[0].endpoints[0].endpoints[0]` to infinity.

I assume this is also the root cause for #12316.

One way to resolve this is to deep copy the endpoint. Since this is a small dict, we can ignore the performance penalty.

```
{{- $endpoint := merge (deepCopy .) $.Values.metrics.serviceMonitor }}
```

Reply:
I see the issue. I also thought about that when adding that. But I tested - at least with `helm template` and there was no issue then.

I'll create a PR to fix this. The `deepCopy` fix is even part of the documentation, as it seems. So  I'll go for that.

Just to understand the issue: Does `template` differ from `upgrade --install` in that case? Or is some specific setup necessary?

Reply:
#12333

Reply:
I haven't confirmed this with tests, so take this with a grain of salt, but this the specific error originates from Helm's storage driver. So when installing the chart, Helm tries to store the final [Release](https://github.com/helm/helm/blob/4b18b19a5e0b11450b9dc92edc75bdd7891c1f4e/pkg/release/release.go#L22) object using the selected storage driver (by default, this is the "secret" storage driver which uses k8s secrets) and at this point, the `Release.Config` (aka the `.Values`) are the ones modified by the rendering process, so they include this reference cycle. I'm not sure whether I consider this a feature or a bug, but I guess that's how it is :D
So when the driver tries to marshal the release object into JSON, the json package throws an error because of this reference cycle:

https://github.com/helm/helm/blob/4b18b19a5e0b11450b9dc92edc75bdd7891c1f4e/pkg/storage/driver/util.go#L36

This error is then caught by the driver which spits out the error message that we see in #12316

https://github.com/helm/helm/blob/a499b4b179307c267bdf3ec49b880e3dbd2a5591/pkg/storage/driver/secrets.go#L154-L157

So yeah, the magic sauce is that Helm needs to try to prepare the stored release, which doesn't happen when using template.

Also, thanks for your swift action :)